rust/projects/file-io/project.md

```diff
@@ -187,22 +236,164 @@ to read about the "get" command. It may help to keep both in mind, or to
+performance and reduced storage. As you implement them, pay attention to what
+exactly they are optimizing for.
+
+As we've described, the database you are building maintains an in-memory index
+of all keys in the database. That index maps from string keys to log pointers,
+not the values themselves.
+
+This change introduces the need to perform reads from the log
+at arbitrary offsets. Consider how that might impact the way
+you manage file handles.
+
+_If, in the previous steps, you elected to store the string values directly in
+memory, now is the time to update your code to store log pointers instead,
+loading from disk on demand._
+
+
+# Part 6: Stateless vs. stateful `KvStore`
+
+Remember that our project is both a library and a command-line program.
+They have sligtly different requirements: the `kvs` CLI commits a single change
+to disk, then exits (it is stateless); the `KvStore` type commits
+changes to disk, then stays resident in memory to service future
+queries (it is stateful).
+
+Is your `KvsStore` stateful or stateless?
+
+_Make your `KvStore` retain the index in memory so it doesn't need to
+re-evaluate it for every call to `get`._
+
+
+
+
+# Part 7: Compacting the log
+
+At this point the database works just fine, but the log grows indefinitely. That
+is appropriate for some databases, but not the one we're building &mdash; we
+want to minimize disk usage as much as we can.
+
+So the final step in creating your database is to compact the log. Consider
+that as the log grows that multiple entries may set the value of a given key:
+
+| idx | command |
+|:---:|:--------|
+| 0 | Command::Set("key-1", "value-1a") |
+| 1 | Command::Set("key-2", "value-2") |
+| | ... |
+| 100 | Command::Set("key-1", "value-1b") |
+
+Consider also that only the most recent command that modified a given key has
+any effect on the current value of that key:
+
+| idx | command |
+|:---:|:--------|
+| 0 | ~Command::Set("key-1", "value-1a")~ |
+| 1 | Command::Set("key-2", "value-2") |
+| | ... |
+| 100 | Command::Set("key-1", "value-1b") |
+
+In this example obviously the command at index 0 is redundant, so it doesn't
+need to be stored. Log compaction then is about rebuilding the log to remove
+redundancy:
+
+| idx | command |
+|:---:|:--------|
+| 0 | Command::Set("key-2", "value-2") |
+| | ... |
+| 99 | Command::Set("key-1", "value-1b") |
+
+Here's the basic algorithm you will use:
+
+- After a commit ...
+- If the _total on-disk size of all un-compacted log entries is greater than 1MB_ ...
+- Re-build the log to remove redundant entries
+
+_How_ you re-build the log is up to you. Consider questions like: what is the
+naive solution? How much memory do you need? What is the minimum amount of
+copying necessary to compact the log? Can the compaction be done in-place? How
+do you maintain data-integrity of compaction fails?
+
+_Implement log compaction for your database._
+
+With compaction, what is the amortized complexity (big O) of calling `set`? What
```

Comment:
I believe the complexity is O(n^2) using a fixed threshold (such as 1MB mentioned above). Are we misleading on purpose?